presto-main/src/main/java/io/prestosql/sql/planner/LocalDynamicFilter.java

```diff
@@ -0,0 +1,211 @@
+ * limitations under the License.
+ */
+package io.prestosql.sql.planner;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Multimap;
+import com.google.common.collect.MultimapBuilder;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.SettableFuture;
+import io.airlift.log.Logger;
+import io.prestosql.spi.predicate.Domain;
+import io.prestosql.spi.predicate.TupleDomain;
+import io.prestosql.sql.DynamicFilters;
+import io.prestosql.sql.planner.optimizations.PlanNodeSearcher;
+import io.prestosql.sql.planner.plan.FilterNode;
+import io.prestosql.sql.planner.plan.JoinNode;
+import io.prestosql.sql.planner.plan.PlanNode;
+import io.prestosql.sql.planner.plan.TableScanNode;
+import io.prestosql.sql.tree.SymbolReference;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Consumer;
+
+import static com.google.common.base.MoreObjects.toStringHelper;
+import static com.google.common.base.Verify.verify;
+import static java.util.Objects.requireNonNull;
+import static java.util.stream.Collectors.toMap;
+
+public class LocalDynamicFilter
+{
+    private static class Channel
+    {
+        private final int buildChannelIndex;
+        private final DynamicFilters.Descriptor descriptor;
+
+        private Channel(int buildChannelIndex, DynamicFilters.Descriptor descriptor)
+        {
+            this.buildChannelIndex = buildChannelIndex;
+            this.descriptor = descriptor;
+        }
+
+        @Override
+        public String toString()
+        {
+            return toStringHelper(this)
+                    .add("buildChannelIndex", buildChannelIndex)
+                    .add("descriptor", descriptor)
+                    .toString();
+        }
+    }
+
+    private static final Logger log = Logger.get(LocalDynamicFilter.class);
+
+    // Mapping from dynamic filter ID to its probe symbols.
+    private final Multimap<String, Symbol> probeSymbols;
+
+    // Mapping from dynamic filter ID to its build channel indices.
+    private final Map<String, Integer> buildChannels;
+
+    private final SettableFuture<TupleDomain<Symbol>> resultFuture;
+    private final Consumer<TupleDomain<String>> tupleDomainConsumer;
+
+    private TupleDomain<String>[] partitions;
+    private int partitionCounter;
+
+    public LocalDynamicFilter(Multimap<String, Symbol> probeSymbols, Map<String, Integer> buildChannels, int partitionCount)
+    {
+        this.probeSymbols = requireNonNull(probeSymbols, "probeSymbols is null");
+        this.buildChannels = requireNonNull(buildChannels, "buildChannels is null");
+        verify(probeSymbols.keySet().equals(buildChannels.keySet()), "probeSymbols and buildChannels must have same keys");
+
+        this.resultFuture = SettableFuture.create();
+        this.tupleDomainConsumer = this::addPartition;
```

Comment:
not needed, inline into `getTupleDomainConsumer`

Reply:
You're right, thanks!